bigc/exceptions.py

```diff
@@ -1,83 +1,181 @@
 from __future__ import annotations
 
-from typing import Optional
+import inspect
+from typing import Any, ClassVar
 
-from requests import Response
+import requests
 
+_status_code_exc_class_map: dict[int, type[BigCommerceException]] = {}
 
-class BigCommerceAPIException(Exception):
-    """Base exception class for BigCommerce API errors"""
 
-    def __init__(self, message: Optional[str] = None, response: Optional[Response] = None):
-        self._message = message
-        self.response = response
+class BigCommerceException(Exception):
+    """Base exception class for BigCommerce API errors."""
 
-        super().__init__(
-            message or f'Request to BigCommerce failed{f" with {self.message}" if response is not None else ""}'
-        )
+    DEFAULT_MESSAGE: ClassVar[str] = 'Request to BigCommerce failed.'
+    STATUS_CODE: ClassVar[int | None] = None
 
-    @property
-    def status_code(self) -> Optional[int]:
-        if self.response is None:
-            return None
+    def __init__(self,
+                 message: str | None = None,
+                 response: Any = None,
+                 errors: dict[str, str] | None = None):
+        """
+        :param message: A short description of the error's cause.
+        :param response: The underlying API response, if any, that caused the
+            error. Currently, this is a ``requests.Response``, but this is an
+            implementation detail that's subject to change.
+        :param errors: A dictionary that may contain additional error details.
+        """
+        super().__init__(message, response, errors)
```

Comment:
```suggestion
        super().__init__(message or self.DEFAULT_MESSAGE, response, errors)
```

?

Reply:
I think you are going to want a way to tell if BigCommerce gave you an error message or not. My thinking is that `exc.message` will give you exactly what BigCommerce gave you (could be `None`), but `str(exc)` is guaranteed to give you an error message, even if it's just the default.

Reply:
Also this would affect pickling. When an exception gets un-pickled, the `args` attribute gets spread into the class' `__init__`. If I made this change, a round-trip pickling of `BigCommerceException(None, None, None)` would result in `BigCommerceException('some default msg', None, None)`.

Reply:
Interesting, ty